src/fsharp/FSharp.Core/async.fs

```diff
@@ -1250,31 +1250,28 @@ namespace Microsoft.FSharp.Control
                                 |> unfake)
                     | Some maxDegreeOfParallelism ->
                         let mutable i = -1
-                        let worker = MakeAsync (fun _ ->
-                            while i < tasks.Length do
+                        let rec worker (trampolineHolder : TrampolineHolder) =
+                            if i < tasks.Length then
                                 let j = Interlocked.Increment &i
                                 if j < tasks.Length then
-                                    let trampolineHolder = new TrampolineHolder()
-                                    trampolineHolder.ExecuteWithTrampoline (fun () ->
-                                        let ctxt =
+                                    if innerCTS.Token.IsCancellationRequested then
+                                        let cexn = new OperationCanceledException (innerCTS.Token)
+                                        recordFailure (Choice2Of2 cexn) |> unfake
+                                        worker trampolineHolder |> unfake
+                                    else
+                                        let taskCtxt =
                                             AsyncActivation.Create
                                                 innerCTS.Token
                                                 trampolineHolder
-                                                (fun res -> recordSuccess j res)
-                                                (fun edi -> recordFailure (Choice1Of2 edi))
-                                                (fun cexn -> recordFailure (Choice2Of2 cexn))
-                                        tasks.[j].Invoke ctxt
-                                    )
-                                    |> unfake
+                                                (fun res -> recordSuccess j res |> unfake; worker trampolineHolder)
+                                                (fun edi -> recordFailure (Choice1Of2 edi) |> unfake; worker trampolineHolder)
+                                                (fun cexn -> recordFailure (Choice2Of2 cexn) |> unfake; worker trampolineHolder)
+                                        tasks.[j].Invoke taskCtxt |> unfake
                             fake()
-                        )
                         for x = 1 to maxDegreeOfParallelism do
-                            QueueAsync
-                                innerCTS.Token
-                                (fun _ -> fake())
-                                (fun edi -> recordFailure (Choice1Of2 edi))
-                                (fun cexn -> recordFailure (Choice2Of2 cexn))
-                                worker
+                            let trampolineHolder = new TrampolineHolder()
```

Comment:
Are there any performance ramifications from allocating one trampoline per worker? I'm not very familiar with the implementation here.

Reply:
QueueAsync allocates one, so Parallel with maxDegreeOfParallelism=None allocates one per computation. So this is no worse.